tchannel/schemes/json.py

```diff
@@ -0,0 +1,53 @@
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals
+)
+
+import json
+
+from tornado import gen
+
+from . import JSON
+
+
+class JsonArgScheme(object):
+    """Semantic params and serialization for json."""
+
+    NAME = JSON
+
+    def __init__(self, tchannel):
+        self._tchannel = tchannel
+
+    @gen.coroutine
+    def __call__(self, service, endpoint, body=None,
+                 headers=None, timeout=None):
+
+        if headers is None:
+            headers = {}
+
+        if body is None:
+            body = {}
+
+        # serialize
+        headers = json.dumps(headers)
+        body = json.dumps(body)
+
+        response = yield self._tchannel.call(
+            scheme=self.NAME,
+            service=service,
+            arg1=endpoint,
+            arg2=headers,
+            arg3=body,
+            timeout=timeout,
+        )
+
+        # deserialize
+        response.headers = json.loads(response.headers)
+        response.body = json.loads(response.body)
```

Comment:
call is the workhorse function. you want it doing pre and post processing too. It doesn't need to how that serialization/deserialization is happening. It just gets an object that does. The default implementation for that object could just be the identity function, so that it doesn't really modify anything (which is what would be used in case of raw because no serialization/deserialization needed).

Also, I would like to avoid mutating attributes in different contexts like that if possible.